From a51cdaeb643e3a4bfed704aaa040f0f700c67d88 Mon Sep 17 00:00:00 2001
From: Omri Itach <omrii@marvell.com>
Date: Tue, 3 Feb 2015 16:45:02 +0200
Subject: [PATCH 0129/1240] armadalp: add preliminary support for Armada LP

	Added preliminary structure support for Armada LP:
	- CPU (armv7 - Cortex A17) , board, and include folders
	- defconfig and SoC defintion files

	* compiled with mvebu_armadalp_defconfig

Signed-off-by: Omri Itach <omrii@marvell.com>
Change-Id: I70cda2f970c08d69fc9dc28beb1cdbabe9beeef1
Reviewed-on: http://vgitil04.il.marvell.com:8080/16431
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Hanna Hawa <hannah@marvell.com>
---
 arch/arm/Kconfig                                |   3 +
 arch/arm/cpu/armv7/Makefile                     |   1 +
 arch/arm/cpu/armv7/armadalp/Kconfig             |  26 ++++++
 arch/arm/cpu/armv7/armadalp/Makefile            |  21 +++++
 arch/arm/cpu/armv7/armadalp/clock.c             |  45 ++++++++++
 arch/arm/cpu/armv7/armadalp/soc-init.c          |  37 ++++++++
 arch/arm/cpu/armv7/armadalp/tables.c            | 114 ++++++++++++++++++++++++
 arch/arm/cpu/mvebu-common/Kconfig               |   1 +
 arch/arm/include/asm/arch-armadalp/armadalp.h   |  26 ++++++
 arch/arm/include/asm/arch-armadalp/memory-map.h |  32 +++++++
 arch/arm/include/asm/arch-armadalp/regs-base.h  |  38 ++++++++
 arch/arm/include/asm/arch-armadalp/soc-info.h   |  39 ++++++++
 board/mvebu/armadalp/Makefile                   |  23 +++++
 board/mvebu/armadalp/armadalp.c                 |  36 ++++++++
 board/mvebu/armadalp/board-info.c               |  91 +++++++++++++++++++
 board/mvebu/armadalp/board-info.h               |  36 ++++++++
 board/mvebu/armadalp/customer-board.c           |  38 ++++++++
 board/mvebu/armadalp/db0-armadalp-info.c        |  38 ++++++++
 board/mvebu/armadalp/db1-armadalp-info.c        |  38 ++++++++
 board/mvebu/common/board.h                      |   2 +-
 configs/mvebu_armadalp_defconfig                |   3 +
 include/configs/armadalp.h                      |  97 ++++++++++++++++++++
 22 files changed, 784 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/cpu/armv7/armadalp/Kconfig
 create mode 100644 arch/arm/cpu/armv7/armadalp/Makefile
 create mode 100644 arch/arm/cpu/armv7/armadalp/clock.c
 create mode 100644 arch/arm/cpu/armv7/armadalp/soc-init.c
 create mode 100644 arch/arm/cpu/armv7/armadalp/tables.c
 create mode 100644 arch/arm/include/asm/arch-armadalp/armadalp.h
 create mode 100644 arch/arm/include/asm/arch-armadalp/memory-map.h
 create mode 100644 arch/arm/include/asm/arch-armadalp/regs-base.h
 create mode 100644 arch/arm/include/asm/arch-armadalp/soc-info.h
 create mode 100644 board/mvebu/armadalp/Makefile
 create mode 100644 board/mvebu/armadalp/armadalp.c
 create mode 100644 board/mvebu/armadalp/board-info.c
 create mode 100644 board/mvebu/armadalp/board-info.h
 create mode 100644 board/mvebu/armadalp/customer-board.c
 create mode 100644 board/mvebu/armadalp/db0-armadalp-info.c
 create mode 100644 board/mvebu/armadalp/db1-armadalp-info.c
 create mode 100644 configs/mvebu_armadalp_defconfig
 create mode 100644 include/configs/armadalp.h

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 5e5a00f..60ef9c8 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -108,6 +108,9 @@ config TARGET_ARMADA_8K
 config TARGET_ARMADA_38X
 	bool "Support Armada 38x"
 
+config TARGET_ARMADA_LP
+	bool "Support Armada LP"
+
 config TARGET_SMDK2410
 	bool "Support smdk2410"
 	select CPU_ARM920T
diff --git a/arch/arm/cpu/armv7/Makefile b/arch/arm/cpu/armv7/Makefile
index 1a9793a..b904828 100644
--- a/arch/arm/cpu/armv7/Makefile
+++ b/arch/arm/cpu/armv7/Makefile
@@ -41,6 +41,7 @@ endif
 obj-$(if $(filter am33xx,$(SOC)),y) += am33xx/
 obj-$(if $(filter armada-xp,$(SOC)),y) += armada-xp/
 obj-$(if $(filter armada38x,$(SOC)),y) += armada38x/
+obj-$(if $(filter armadalp,$(SOC)),y) += armadalp/
 obj-$(CONFIG_AT91FAMILY) += at91/
 obj-$(if $(filter bcm281xx,$(SOC)),y) += bcm281xx/
 obj-$(if $(filter bcmcygnus,$(SOC)),y) += bcmcygnus/
diff --git a/arch/arm/cpu/armv7/armadalp/Kconfig b/arch/arm/cpu/armv7/armadalp/Kconfig
new file mode 100644
index 0000000..7fbd296
--- /dev/null
+++ b/arch/arm/cpu/armv7/armadalp/Kconfig
@@ -0,0 +1,26 @@
+if TARGET_ARMADA_LP
+
+config SYS_CPU
+	string
+	default "armv7"
+
+config SYS_BOARD
+	string
+	default "armadalp"
+
+config SYS_VENDOR
+	string
+	default "mvebu"
+
+config SYS_SOC
+	string
+	default "armadalp"
+
+config SYS_CONFIG_NAME
+	string
+	default "armadalp"
+
+config PALLADIUM
+	bool "Support palladium emulation"
+	default y
+endif
diff --git a/arch/arm/cpu/armv7/armadalp/Makefile b/arch/arm/cpu/armv7/armadalp/Makefile
new file mode 100644
index 0000000..c99db44
--- /dev/null
+++ b/arch/arm/cpu/armv7/armadalp/Makefile
@@ -0,0 +1,21 @@
+#
+# ***************************************************************************
+# Copyright (C) 2015 Marvell International Ltd.
+# ***************************************************************************
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the Free
+# Software Foundation, either version 2 of the License, or any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# ***************************************************************************
+#
+
+obj-y += soc-init.o
+obj-y += tables.o
+obj-y += clock.o
diff --git a/arch/arm/cpu/armv7/armadalp/clock.c b/arch/arm/cpu/armv7/armadalp/clock.c
new file mode 100644
index 0000000..7d66f49
--- /dev/null
+++ b/arch/arm/cpu/armv7/armadalp/clock.c
@@ -0,0 +1,45 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#include <common.h>
+#include <asm/arch-mvebu/clock.h>
+
+u32 soc_tclk_get(void)
+{
+	return 200000000;
+}
+
+u32 soc_cpu_clk_get(void)
+{
+	return 200000000;
+}
+
+u32 soc_ddr_clk_get(void)
+{
+	return 200000000;
+}
+
+u32 soc_l2_clk_get(void)
+{
+	return 800000000;
+}
+
+u32 soc_timer_clk_get(void)
+{
+	return 800000000;
+}
diff --git a/arch/arm/cpu/armv7/armadalp/soc-init.c b/arch/arm/cpu/armv7/armadalp/soc-init.c
new file mode 100644
index 0000000..70d9314
--- /dev/null
+++ b/arch/arm/cpu/armv7/armadalp/soc-init.c
@@ -0,0 +1,37 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#define DEBUG
+
+#include <common.h>
+#include <asm/io.h>
+#include <asm/arch-mvebu/soc.h>
+#include <asm/arch-mvebu/unit-info.h>
+#include <asm/arch-armadalp/armadalp.h>
+
+int soc_early_init_f(void)
+{
+	return 0;
+}
+
+struct mvebu_soc_family *soc_init(void)
+{
+	/* Do early SOC specific init here */
+
+	return &armadalp_family_info;
+}
diff --git a/arch/arm/cpu/armv7/armadalp/tables.c b/arch/arm/cpu/armv7/armadalp/tables.c
new file mode 100644
index 0000000..180f727
--- /dev/null
+++ b/arch/arm/cpu/armv7/armadalp/tables.c
@@ -0,0 +1,114 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#include <common.h>
+#include <asm/io.h>
+#include <asm/arch/regs-base.h>
+#include <asm/arch-mvebu/adec.h>
+#include <asm/arch-mvebu/unit-info.h>
+#include <asm/arch-mvebu/soc.h>
+#include <asm/arch-mvebu/mpp.h>
+
+struct adec_win memory_map[] = {
+#ifndef CONFIG_PALLADIUM
+	{PEX_0_TARGET_ID,	MVEBU_PCIE_MEM_BASE(0),	MVEBU_PCIE_MEM_SIZE(0), PCI_0_MEM_ATTR, 1, 1},
+	{DEV_TARGET_ID,		NOR_CS_BASE,		NOR_CS_SIZE,	 BOOT_CS_ATTR,   0, 1},
+#endif
+	{TBL_TERM, TBL_TERM, TBL_TERM, TBL_TERM, TBL_TERM, TBL_TERM},
+};
+
+/* List the maximum amount of units for all devices and boards */
+u16 base_unit_mask_table[MAX_UNIT_ID] = {
+	/* the following entries are temporary examples only -  referenced from A38x */
+	[PCIE_UNIT_ID]		=  cnt_to_mask(1),
+	[SATA_UNIT_ID]		=  cnt_to_mask(1),
+	[ETH_GIG_UNIT_ID]	=  cnt_to_mask(2),
+	[UART_UNIT_ID]		=  cnt_to_mask(1),
+	[USB_UNIT_ID]		=  cnt_to_mask(1),
+	[USB3_UNIT_ID]		=  cnt_to_mask(1),
+	[SPI_UNIT_ID]		=  cnt_to_mask(1),
+	[XOR_UNIT_ID]		=  cnt_to_mask(1),
+	[SDIO_UNIT_ID]		=  cnt_to_mask(1),
+	[NAND_UNIT_ID]		=  cnt_to_mask(1)
+};
+
+/* Per device unit info. List which units are disabled for this device */
+u16 armadalp_A_unit_disable[MAX_UNIT_ID] = {
+	[ETH_GIG_UNIT_ID]	=  id_to_mask(2)
+};
+
+u16 armadalp_B_unit_disable[MAX_UNIT_ID] = {
+	[ETH_GIG_UNIT_ID]	= id_to_mask(1),
+	[USB3_UNIT_ID]		= id_to_mask(1)
+};
+
+/* MPP description table. Informative only */
+char *armadalp_mpp_desc[MAX_MPP_ID + 1][MAX_MPP_OPTS] = {
+/*	0x0        0x1         0x2              0x3             0x4             0x5             0x6 */
+	/* the following entries are temporary examples only -  referenced from A38x */
+/* 0 */	{"GPIO_0", "NF_IO_2", "SPI_0_CSn",	"SPI_1_CSn",	"NA",           "ND_IO_2", "NA"},	/* 0  */
+/* 1 */	{"GPIO_1", "NF_IO_3", "SPI_0_MOSI",	"SPI_1_MOSI",	"NA",	        "ND_IO_3", "NA"},	 /* 1  */
+/* 2 */	{"GPIO_2", "NF_IO_4", "PTP_EVENT_REQ",  "C0_LED",	"AU_I2S_DI",	"ND_IO_4", "SPI_1_MOSI"}, /* 2  */
+/* 3 */	{"GPIO_3", "NF_IO_5", "PTP_TRIG_GEN",	"P3_LED",	"AU_I2S_MCLK",	"ND_IO_5", "SPI_1_MISO"}, /* 1  */
+};
+
+/* Describe the available MPP buses */
+struct mpp_bus armadalp_mpp_buses[MAX_MPP_BUS] = {
+	/* the following entries are temporary examples only -  referenced from A38x */
+	[UART_0_MPP_BUS] = {"UART 0", 2, 1, {{{0, 0x1 }, {0, 0x1} } } },
+	[UART_1_MPP_BUS] = {"UART 1", 2, 2, {{{4, 0x2}, {5, 0x2} }, {{19, 0x6}, {20, 0x6} } } },
+	[I2C_0_MPP_BUS] = {"I2C 0", 2, 1, {{{2, 0x1}, {3, 0x1} } } },
+	[GE_0_MPP_BUS] = {"GE 0", 14, 1, {{{4, 0x1}, {5, 0x1}, {6, 0x1}, {7, 0x1}, {8, 0x1}, {9, 0x1}, {10, 0x1}, {11, 0x1}, \
+										{12, 0x1}, {13, 0x1}, {14, 0x1}, {15, 0x1}, {16, 0x1}, {17, 0x1} } } },
+	[GE_1_MPP_BUS] = {"GE 1", 12, 1, {{{21, 0x2}, {27, 0x2}, {28, 0x2}, {29, 0x2}, {30, 0x2}, {31, 0x2}, {32, 0x2}, {37, 0x2}, {38, 0x2}, {39, 0x2}, {40, 0x2}, {41, 0x2} } } },
+	[SPI_0_MPP_BUS]  = {"SPI 0", 4, 1, {{{22, 0x1}, {23, 0x1}, {24, 0x1}, {25, 0x1} } } },
+	[M_VTT_0_MPP_BUS]  = {"M_VTT_CTRL", 1, 1, {{{43, 0x2} } } },
+	[SDIO_0_MPP_BUS]  = {"SDIO 0", 10, 1, {{{48, 0x5}, {49, 0x5}, {50, 0x5}, {52, 0x5}, {53, 0x5}, {54, 0x5}, {55, 0x5}, {57, 0x5}, {58, 0x5}, {59, 0x5} } } },
+	[NAND_0_MPP_BUS] = {"NAND 0", 14, 1, {{{22, 0x5}, {23, 0x5}, {25, 0x5}, {28, 0x5}, {30, 0x5}, {32, 0x5},
+						{33, 0x5}, {34, 0x5}, {35, 0x5}, {36, 0x5}, {38, 0x5}, {40, 0x5},
+						{41, 0x6}, {42, 0x5} } } },
+};
+
+struct mvebu_soc_info armadalp_soc_info[] = {
+	/* Armada LP A */
+	{
+		.name = "MV-ARMADALP_A", /* temp SOC name and SOC ID */
+		.id = 0x9991,
+		.unit_disable = armadalp_A_unit_disable,
+		.memory_map = memory_map
+	},
+	/* Armada LP B */
+	{
+		.name = "MV-ARMADALP_B", /* temp SOC name and SOC ID */
+		.id = 0x9992,
+		.unit_disable = armadalp_B_unit_disable,
+		.memory_map = memory_map
+	},
+	/* Delimiter */
+	{.id = 0}
+};
+
+struct mvebu_soc_family armadalp_family_info = {
+	.name = "MV-99xx", /* temp SOC name and SOC ID */
+	.id   = 9990,
+	.rev_name = {"Z1", "A0"},
+	.base_unit_info = base_unit_mask_table,
+	.soc_table = armadalp_soc_info,
+	.mpp_desc  = (char **)armadalp_mpp_desc,
+	.mpp_buses = armadalp_mpp_buses,
+};
diff --git a/arch/arm/cpu/mvebu-common/Kconfig b/arch/arm/cpu/mvebu-common/Kconfig
index 29e9a42..4397ff0 100644
--- a/arch/arm/cpu/mvebu-common/Kconfig
+++ b/arch/arm/cpu/mvebu-common/Kconfig
@@ -1,6 +1,7 @@
 
 source "arch/arm/cpu/armv8/armada8k/Kconfig"
 source "arch/arm/cpu/armv7/armada38x/Kconfig"
+source "arch/arm/cpu/armv7/armadalp/Kconfig"
 
 config MVEBU
 	bool "Support MVEBU devices"
diff --git a/arch/arm/include/asm/arch-armadalp/armadalp.h b/arch/arm/include/asm/arch-armadalp/armadalp.h
new file mode 100644
index 0000000..5a769c8
--- /dev/null
+++ b/arch/arm/include/asm/arch-armadalp/armadalp.h
@@ -0,0 +1,26 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#ifndef _ARMADALP_H_
+#define _ARMADALP_H_
+
+#include <asm/arch-mvebu/mvebu.h>
+
+extern struct mvebu_soc_family armadalp_family_info;
+
+#endif	/* _ARMADA8K_H_ */
diff --git a/arch/arm/include/asm/arch-armadalp/memory-map.h b/arch/arm/include/asm/arch-armadalp/memory-map.h
new file mode 100644
index 0000000..7da88f7
--- /dev/null
+++ b/arch/arm/include/asm/arch-armadalp/memory-map.h
@@ -0,0 +1,32 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+/* Memory Map */
+/* 0x00000000 - 0xE0000000 - DRAM */
+/* 0xE0000000 - 0xE2000000 - PCI-0 */
+/* 0xE2000000 - 0xE4000000 - PCI-1 */
+/* 0xE4000000 - 0xE6000000 - PCI-2 */
+/* 0xE6000000 - 0xE8000000 - PCI-3 */
+/* 0xF1000000 - 0xF1100000 - INTERNAL_REG */
+
+#define MVEBU_REGS_BASE			(0xF1000000)
+#define MVEBU_REGS_SIZE			(0x100000)
+
+#define MVEBU_PCIE_MEM_BASE(id)		(0xE0000000 + (id * 0x2000000))
+#define MVEBU_PCIE_MEM_SIZE(id)		(0x2000000)
+
diff --git a/arch/arm/include/asm/arch-armadalp/regs-base.h b/arch/arm/include/asm/arch-armadalp/regs-base.h
new file mode 100644
index 0000000..ad40360
--- /dev/null
+++ b/arch/arm/include/asm/arch-armadalp/regs-base.h
@@ -0,0 +1,38 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#ifndef _REGS_BASE_H_
+#define _REGS_BASE_H_
+
+#include <asm/arch/memory-map.h>
+
+#define MPP_REGS_BASE		(MVEBU_REGS_BASE + 0x18000)
+#define MVEBU_ADEC_BASE		(MVEBU_REGS_BASE + 0x20000)
+
+#define MVEBU_UART_BASE(x)	(MVEBU_REGS_BASE + 0x12000 + (0x100 * x))
+#define MVEBU_GLOBAL_TIMER_BASE	(MVEBU_REGS_BASE + 0x20300)
+#define MVEBU_PCIE_BASE(x)	((x == 0) ? (MVEBU_REGS_BASE + 0x80000) : \
+				(MVEBU_REGS_BASE + 0x40000 + (0x4000 * (x - 1))))
+
+#define MVEBU_MISC_REGS_BASE	(MVEBU_REGS_BASE + 0x18200)
+#define MVEBU_DEVICE_ID_REG	(MVEBU_MISC_REGS_BASE + 0x38)
+#define MVEBU_DEVICE_REV_REG	(MVEBU_MISC_REGS_BASE + 0x3C)
+#define MVEBU_RESET_MASK_REG	(MVEBU_MISC_REGS_BASE + 0x60)
+#define MVEBU_SOFT_RESET_REG	(MVEBU_MISC_REGS_BASE + 0x64)
+
+#endif	/* _REGS_BASE_H_ */
diff --git a/arch/arm/include/asm/arch-armadalp/soc-info.h b/arch/arm/include/asm/arch-armadalp/soc-info.h
new file mode 100644
index 0000000..54bea6c
--- /dev/null
+++ b/arch/arm/include/asm/arch-armadalp/soc-info.h
@@ -0,0 +1,39 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#ifndef _SOC_INFO_H_
+#define _SOC_INFO_H_
+
+/* MBUS definitions*/
+#define MAX_MBUS_WINS			10
+#define MAX_MBUS_REMAP_WINS		8
+#define INTERNAL_REG_WIN_NUM		11
+
+/* General MPP definitions */
+#define MAX_MPP_OPTS		7
+#define MAX_MPP_ID		56
+
+#define MPP_BIT_CNT		4
+#define MPP_FIELD_MASK		0x7
+#define MPP_FIELD_BITS		3
+#define MPP_VAL_MASK		0xF
+
+#define MPPS_PER_REG		(32 / MPP_BIT_CNT)
+#define MAX_MPP_REGS		((MAX_MPP_ID + MPPS_PER_REG) / MPPS_PER_REG)
+
+#endif	/* _SOC_INFO_H_ */
diff --git a/board/mvebu/armadalp/Makefile b/board/mvebu/armadalp/Makefile
new file mode 100644
index 0000000..c7d525a
--- /dev/null
+++ b/board/mvebu/armadalp/Makefile
@@ -0,0 +1,23 @@
+#
+# ***************************************************************************
+# Copyright (C) 2015 Marvell International Ltd.
+# ***************************************************************************
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the Free
+# Software Foundation, either version 2 of the License, or any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# ***************************************************************************
+#
+
+obj-y	+= armadalp.o
+obj-y	+= board-info.o
+obj-y	+= db0-armadalp-info.o
+obj-y	+= db1-armadalp-info.o
+obj-y	+= customer-board.o
diff --git a/board/mvebu/armadalp/armadalp.c b/board/mvebu/armadalp/armadalp.c
new file mode 100644
index 0000000..2fbeada
--- /dev/null
+++ b/board/mvebu/armadalp/armadalp.c
@@ -0,0 +1,36 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+/* #define DEBUG*/
+
+#include <common.h>
+#include <malloc.h>
+#include <errno.h>
+#include <netdev.h>
+#include <asm/io.h>
+#include <asm/arch-armadalp/armadalp.h>
+#include <linux/compiler.h>
+#include "board-info.h"
+
+
+struct mvebu_board_family *board_init_family(void)
+{
+	/* Board family specific init goes here */
+
+	return &armadalp_board_family;
+}
diff --git a/board/mvebu/armadalp/board-info.c b/board/mvebu/armadalp/board-info.c
new file mode 100644
index 0000000..e3a2bd1
--- /dev/null
+++ b/board/mvebu/armadalp/board-info.c
@@ -0,0 +1,91 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#include <common.h>
+#include <linux/compiler.h>
+#include <asm/arch-mvebu/unit-info.h>
+#include "board-info.h"
+
+#define SATR_EEPROM0_ADDR	0x22
+
+/* Define all SAR variables available for SOC */
+/* Assuming their location is equal on all boards */
+struct sar_var armadalp_sar_lookup[MAX_SAR] = {
+	/* the following entries are temporary examples only -  referenced from A38x */
+	[CPUS_NUM_SAR] = {0, 2, 3, 0,
+		{{0x0, "Single CPU", 0},
+		 {0x2, "Dual CPU", 0},
+		 {0x3, "Quad CPU", VAR_IS_DEFUALT} },
+	},
+	[FREQ_SAR] = {2, 5, 2, 0,
+		{{0x0, "800 / 400 / 400"},
+		 {0x5, "1200 / 600 / 600"} },
+	},
+	[BOOT_SRC_SAR] = {7, 2, 3, 0,
+		{{0x0, "NAND boot", 0},
+		 {0x2, "SPI boot", VAR_IS_DEFUALT},
+		 {0x3, "NOR boot", 0} },
+	}
+};
+
+/* Define general SAR information */
+struct sar_data armadalp_sar = {
+	.chip_addr    = {0x4c, 0x4d, 0x4e, 0x4f},
+	.chip_count   = 4,
+	.bit_width    = 5,
+	.sar_lookup   = armadalp_sar_lookup
+};
+
+/* Define all board configuration variables */
+/* Assuming their location is equal on all boards */
+struct cfg_var armadalp_cfg_lookup[MAX_CFG] = {
+	[BOARD_ID_CFG] = {0, 0, 3, 4, 0,
+		{{ARMADA_LP_DB0_ID, "DB-9990", VAR_IS_DEFUALT},
+		 {ARMADA_LP_DB1_ID, "DB-9991", 0} },
+	},
+	[SERDES_0_CFG] = {1, 0, 3, 2, 0,
+		{{0x0, "pcie 0", VAR_IS_DEFUALT},
+		 {0x1, "sata 0", 0} },
+	},
+	[SERDES_1_CFG] = {1, 3, 3, 3, 0,
+		{{0x0, "pcie 1", 0},
+		 {0x1, "sata 1", VAR_IS_DEFUALT},
+		 {0x2, "usb3 1", 0} },
+	}
+};
+
+/* Define general SAR information */
+struct cfg_data armadalp_cfg = {
+	.chip_addr    = 0x55,
+	.cfg_lookup   = armadalp_cfg_lookup
+};
+
+
+
+struct mvebu_board_info *armadalp_board_lookup[MAX_BOARD_ID] = {
+	[ARMADA_LP_DB0_ID] = &db0_armadalp_info,
+	[ARMADA_LP_DB1_ID] = &db1_armadalp_info
+};
+struct mvebu_board_family armadalp_board_family = {
+	.default_id = ARMADA_LP_DB0_ID,
+	.board_cnt = MAX_BOARD_ID,
+	.boards_info = armadalp_board_lookup,
+	.sar = &armadalp_sar,
+	.cfg = &armadalp_cfg,
+};
+
diff --git a/board/mvebu/armadalp/board-info.h b/board/mvebu/armadalp/board-info.h
new file mode 100644
index 0000000..ece8b04
--- /dev/null
+++ b/board/mvebu/armadalp/board-info.h
@@ -0,0 +1,36 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#ifndef _BOARD_INFO_H_
+#define _BOARD_INFO_H_
+
+#include "../common/board.h"
+#include "../common/sar.h"
+
+enum a8k_board_types {
+	ARMADA_LP_DB0_ID = 0,
+	ARMADA_LP_DB1_ID,
+	ARMADA_LP_CUSTOMER0_ID,
+	MAX_BOARD_ID
+};
+
+extern struct mvebu_board_info db0_armadalp_info;
+extern struct mvebu_board_info db1_armadalp_info;
+extern struct mvebu_board_info armadalp_customer_info;
+extern struct mvebu_board_family armadalp_board_family;
+#endif /* _BOARD_INFO_H_ */
diff --git a/board/mvebu/armadalp/customer-board.c b/board/mvebu/armadalp/customer-board.c
new file mode 100644
index 0000000..f9199ac
--- /dev/null
+++ b/board/mvebu/armadalp/customer-board.c
@@ -0,0 +1,38 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#include <common.h>
+#include <linux/compiler.h>
+#include <asm/arch-mvebu/unit-info.h>
+#include "board-info.h"
+
+/* Optionaly define which units to disable */
+u16 armadalp_customer_unit_disable[MAX_UNIT_ID] = {
+	[NAND_UNIT_ID] = id_to_mask(0),
+	[SDIO_UNIT_ID] = id_to_mask(0)
+};
+
+struct mvebu_board_info armadalp_customer_info = {
+	.name = "CUSTOMER-BOARD0-MV-ARMADA-LP",
+	.id = ARMADA_LP_CUSTOMER0_ID,
+	.unit_mask = armadalp_customer_unit_disable,
+	.unit_update_mode = UNIT_INFO_DISABLE,
+	.mpp_regs = {0x11111111, 0x00100001},
+	.configurable = 0,
+};
+
diff --git a/board/mvebu/armadalp/db0-armadalp-info.c b/board/mvebu/armadalp/db0-armadalp-info.c
new file mode 100644
index 0000000..4afa3a1
--- /dev/null
+++ b/board/mvebu/armadalp/db0-armadalp-info.c
@@ -0,0 +1,38 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#include <common.h>
+#include <linux/compiler.h>
+#include <asm/arch-mvebu/unit-info.h>
+#include "board-info.h"
+
+struct mvebu_board_config db0_armadalp_config = {
+	.sar_cnt = 2,
+	.active_sar = {BOOT_SRC_SAR, CPUS_NUM_SAR},
+	.cfg_cnt = 2,
+	.active_cfg = {BOARD_ID_CFG, SERDES_0_CFG}
+};
+
+struct mvebu_board_info db0_armadalp_info = {
+	.name = "DB0-MV-ARMADALP",
+	.id = ARMADA_LP_DB0_ID,
+	.mpp_regs = {0x11111111, 0x00100001},
+	.configurable = 1,
+	.config_data = &db0_armadalp_config,
+};
+
diff --git a/board/mvebu/armadalp/db1-armadalp-info.c b/board/mvebu/armadalp/db1-armadalp-info.c
new file mode 100644
index 0000000..1bb6e6a
--- /dev/null
+++ b/board/mvebu/armadalp/db1-armadalp-info.c
@@ -0,0 +1,38 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#include <common.h>
+#include <linux/compiler.h>
+#include <asm/arch-mvebu/unit-info.h>
+#include "board-info.h"
+
+struct mvebu_board_config db1_armadalp_config = {
+	.sar_cnt = 2,
+	.active_sar = {CPUS_NUM_SAR},
+	.cfg_cnt = 2,
+	.active_cfg = {BOARD_ID_CFG}
+};
+
+struct mvebu_board_info db1_armadalp_info = {
+	.name = "DB1-MV-ARMADALP",
+	.id = ARMADA_LP_DB1_ID,
+	.mpp_regs = {0x11111111, 0x00100001},
+	.configurable = 1,
+	.config_data = &db1_armadalp_config,
+};
+
diff --git a/board/mvebu/common/board.h b/board/mvebu/common/board.h
index 921f2b9..857823b 100644
--- a/board/mvebu/common/board.h
+++ b/board/mvebu/common/board.h
@@ -28,7 +28,7 @@
 #include "sar.h"
 #include "config.h"
 
-#define MAX_BOARD_NAME		16
+#define MAX_BOARD_NAME		32
 
 /*
  * mvebu_board_config  - Describes board configuration features
diff --git a/configs/mvebu_armadalp_defconfig b/configs/mvebu_armadalp_defconfig
new file mode 100644
index 0000000..82beec4
--- /dev/null
+++ b/configs/mvebu_armadalp_defconfig
@@ -0,0 +1,3 @@
+CONFIG_ARM=y
+CONFIG_TARGET_ARMADA_LP=y
+CONFIG_MVEBU_UART=y
diff --git a/include/configs/armadalp.h b/include/configs/armadalp.h
new file mode 100644
index 0000000..145fa86
--- /dev/null
+++ b/include/configs/armadalp.h
@@ -0,0 +1,97 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2015 Marvell International Ltd.
+* ***************************************************************************
+* This program is free software: you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the Free
+* Software Foundation, either version 2 of the License, or any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+* ***************************************************************************
+*/
+
+#ifndef __ARMADA_LP_H
+#define __ARMADA_LP_H
+
+/* Version */
+#define CONFIG_IDENT_STRING		" Marvell version: 2015_T2.0"
+#define CONFIG_BOOTP_VCI_STRING		" U-boot.armv7.armadalp"
+
+/* Configurable boards */
+#define CONFIG_MVEBU_DEVEL_BOARD      /* Turn this off for Customer boards */
+
+/* Platform */
+#define CONFIG_MVEBU_CA9
+#define CONFIG_MVEBU_TIMER
+
+/*
+#define MV_INCLUDE_SDIO
+#define MV_INCLUDE_INTEG_SATA
+#define MV_INCLUDE_CLK_PWR_CNTRL
+
+#define MV_INCLUDE_RCVR
+#define MV_INCLUDE_NAND
+#define MV_INCLUDE_SPI
+#define MV_INCLUDE_XOR
+#define MV_INCLUDE_SATA
+#define MV_INCLUDE_USB
+#define CONFIG_MV_ETH_NETA
+#define MV_INCLUDE_GPIO
+#define MV_INCLUDE_I2C
+#define MV_INCLUDE_RTC
+#define MV_INCLUDE_PEX
+*/
+
+/* Plaform */
+#define CONFIG_MARVELL
+
+/* Clock */
+#ifndef __ASSEMBLY__
+	#define CONFIG_SYS_HZ                   1000
+	#define CONFIG_SYS_TCLK                 soc_tclk_get()
+	#define CONFIG_SYS_BUS_CLK              soc_sys_clk_get()
+#endif /* __ASSEMBLY__ */
+
+/* DRAM detection stuff */
+#define MV_DRAM_AUTO_SIZE
+#define CONFIG_SYS_DRAM_BANKS           1
+
+/*
+ * Include the generic MVEBU configuration file.
+ * The generic file is affected by the defines above.
+ */
+#include "mvebu-common.h"
+
+/*
+ * SOC specific overrides
+ * Add here all config values that differ
+ * from the generic value in mvebu-common.h
+ */
+#ifdef CONFIG_PALLADIUM
+
+/* Disable I-CACHE for palladium */
+#define CONFIG_SYS_ICACHE_OFF
+
+#undef CONFIG_BOOTDELAY
+#define CONFIG_BOOTDELAY                -1
+
+#undef CONFIG_BAUDRATE
+#define CONFIG_BAUDRATE                 4000
+
+#undef CONFIG_BOOTARGS
+#define CONFIG_BOOTARGS			"console=ttyS0,5652 earlycon=uart8250,mmio32,0xf0512000 mem=128M"
+
+#undef  CONFIG_EXTRA_ENV_SETTINGS
+#define CONFIG_EXTRA_ENV_SETTINGS	"kernel_addr=0x107ffc0\0"	\
+					"fdt_addr=0x1000000\0"		\
+					"fdt_high=0xa0000000\0"
+#endif /*CONFIG_PALLADIUM*/
+
+
+#endif /* __ARMADA_LP_H */
-- 
1.9.1

